File: backend/app/schemas/report_subscription.py
"""
Report Subscriptions Schemas

Pydantic schemas for report subscription management including
subscriptions, notification preferences, and digest configurations.
"""

from pydantic import BaseModel, Field
from typing import Optional, Any
from enum import Enum
from datetime import datetime


# Enums

class SubscriptionType(str, Enum):
    INSTANT = "instant"  # Notify immediately on changes
    DAILY_DIGEST = "daily_digest"
    WEEKLY_DIGEST = "weekly_digest"
    MONTHLY_DIGEST = "monthly_digest"
    SCHEDULED = "scheduled"  # At specific times
    ON_REFRESH = "on_refresh"  # When data refreshes
    ON_THRESHOLD = "on_threshold"  # When metrics cross thresholds


class SubscriptionStatus(str, Enum):
    ACTIVE = "active"
    PAUSED = "paused"
    EXPIRED = "expired"
    CANCELLED = "cancelled"


class ResourceType(str, Enum):
    DASHBOARD = "dashboard"
    CHART = "chart"
    REPORT = "report"
    KPI = "kpi"
    QUERY = "query"
    ALERT = "alert"


class NotificationChannel(str, Enum):
    EMAIL = "email"
    SLACK = "slack"
    TEAMS = "teams"
    PUSH = "push"
    IN_APP = "in_app"
    SMS = "sms"
    WEBHOOK = "webhook"


class DigestFrequency(str, Enum):
    DAILY = "daily"
    WEEKLY = "weekly"
    BIWEEKLY = "biweekly"
    MONTHLY = "monthly"


class ThresholdOperator(str, Enum):
    GREATER_THAN = "gt"
    GREATER_THAN_OR_EQUAL = "gte"
    LESS_THAN = "lt"
    LESS_THAN_OR_EQUAL = "lte"
    EQUAL = "eq"
    NOT_EQUAL = "neq"
    CHANGE_BY = "change_by"
    CHANGE_BY_PERCENT = "change_by_percent"


# Notification Configuration Schemas

class EmailNotificationConfig(BaseModel):
    email_address: Optional[str] = None  # If different from user's email
    include_preview: bool = True
    include_attachment: bool = False
    attachment_format: str = "pdf"
    subject_prefix: str = "[DataViz]"


class SlackNotificationConfig(BaseModel):
    channel_id: Optional[str] = None
    dm_user: bool = True
    include_preview_image: bool = True
    mention_on_threshold: bool = False


class TeamsNotificationConfig(BaseModel):
    channel_webhook: Optional[str] = None
    include_preview_image: bool = True


class PushNotificationConfig(BaseModel):
    send_to_mobile: bool = True
    send_to_desktop: bool = True
    sound_enabled: bool = True


class WebhookNotificationConfig(BaseModel):
    url: str
    method: str = "POST"
    headers: dict[str, str] = {}
    include_data: bool = True


class NotificationConfig(BaseModel):
    channel: NotificationChannel
    email_config: Optional[EmailNotificationConfig] = None
    slack_config: Optional[SlackNotificationConfig] = None
    teams_config: Optional[TeamsNotificationConfig] = None
    push_config: Optional[PushNotificationConfig] = None
    webhook_config: Optional[WebhookNotificationConfig] = None
    enabled: bool = True


# Threshold Configuration

class ThresholdCondition(BaseModel):
    metric_name: str
    operator: ThresholdOperator
    value: float
    secondary_value: Optional[float] = None  # For range conditions
    comparison_period: Optional[str] = None  # "previous_day", "previous_week", etc.


# Schedule Configuration

class SubscriptionSchedule(BaseModel):
    days_of_week: list[str] = []  # ["monday", "wednesday", "friday"]
    time_of_day: str = "09:00"  # HH:MM format
    timezone: str = "UTC"


# Digest Configuration

class DigestConfig(BaseModel):
    frequency: DigestFrequency
    day_of_week: Optional[str] = None  # For weekly digests
    day_of_month: Optional[int] = None  # For monthly digests
    time_of_day: str = "09:00"
    timezone: str = "UTC"
    include_summary: bool = True
    include_highlights: bool = True
    max_items: int = 10
    group_by_resource: bool = True


# Subscription Schemas

class SubscriptionBase(BaseModel):
    resource_type: ResourceType
    resource_id: str
    resource_name: Optional[str] = None
    subscription_type: SubscriptionType
    notification_channels: list[NotificationConfig]


class SubscriptionCreate(SubscriptionBase):
    schedule: Optional[SubscriptionSchedule] = None
    digest_config: Optional[DigestConfig] = None
    threshold_conditions: list[ThresholdCondition] = []
    filters: dict[str, Any] = {}
    is_active: bool = True
    expires_at: Optional[datetime] = None


class SubscriptionUpdate(BaseModel):
    resource_name: Optional[str] = None
    subscription_type: Optional[SubscriptionType] = None
    notification_channels: Optional[list[NotificationConfig]] = None
    schedule: Optional[SubscriptionSchedule] = None
    digest_config: Optional[DigestConfig] = None
    threshold_conditions: Optional[list[ThresholdCondition]] = None
    filters: Optional[dict[str, Any]] = None
    is_active: Optional[bool] = None
    status: Optional[SubscriptionStatus] = None
    expires_at: Optional[datetime] = None


class Subscription(SubscriptionBase):
    id: str
    user_id: str
    organization_id: Optional[str] = None
    status: SubscriptionStatus = SubscriptionStatus.ACTIVE
    is_active: bool = True
    schedule: Optional[SubscriptionSchedule] = None
    digest_config: Optional[DigestConfig] = None
    threshold_conditions: list[ThresholdCondition] = []
    filters: dict[str, Any] = {}
    last_notified_at: Optional[datetime] = None
    notification_count: int = 0
    expires_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class SubscriptionListResponse(BaseModel):
    subscriptions: list[Subscription]
    total: int


# Notification History Schemas

class NotificationDelivery(BaseModel):
    channel: NotificationChannel
    success: bool
    delivered_at: Optional[datetime] = None
    error: Optional[str] = None


class SubscriptionNotification(BaseModel):
    id: str
    subscription_id: str
    user_id: str
    resource_type: ResourceType
    resource_id: str
    resource_name: Optional[str] = None
    notification_type: str  # "instant", "digest", "threshold", etc.
    title: str
    message: str
    data: dict[str, Any] = {}
    deliveries: list[NotificationDelivery] = []
    read: bool = False
    read_at: Optional[datetime] = None
    clicked: bool = False
    clicked_at: Optional[datetime] = None
    created_at: datetime

    class Config:
        from_attributes = True


class NotificationListResponse(BaseModel):
    notifications: list[SubscriptionNotification]
    total: int
    unread_count: int


# Digest Schemas

class DigestItem(BaseModel):
    resource_type: ResourceType
    resource_id: str
    resource_name: str
    summary: str
    changes: list[str] = []
    metrics: dict[str, Any] = {}
    timestamp: datetime


class Digest(BaseModel):
    id: str
    user_id: str
    digest_type: DigestFrequency
    period_start: datetime
    period_end: datetime
    items: list[DigestItem]
    total_items: int
    highlights: list[str] = []
    sent_at: Optional[datetime] = None
    created_at: datetime

    class Config:
        from_attributes = True


class DigestListResponse(BaseModel):
    digests: list[Digest]
    total: int


# User Preferences Schemas

class GlobalNotificationPreferences(BaseModel):
    email_enabled: bool = True
    push_enabled: bool = True
    in_app_enabled: bool = True
    slack_enabled: bool = False
    teams_enabled: bool = False
    quiet_hours_enabled: bool = False
    quiet_hours_start: str = "22:00"
    quiet_hours_end: str = "08:00"
    timezone: str = "UTC"
    digest_enabled: bool = True
    digest_frequency: DigestFrequency = DigestFrequency.DAILY
    digest_time: str = "09:00"
    unsubscribe_from_marketing: bool = False


class UserSubscriptionPreferences(BaseModel):
    user_id: str
    global_preferences: GlobalNotificationPreferences
    channel_preferences: dict[str, bool] = {}  # Per-channel overrides
    resource_preferences: dict[str, bool] = {}  # Per-resource-type overrides
    muted_resources: list[str] = []  # Resource IDs to never notify about
    updated_at: datetime

    class Config:
        from_attributes = True


# Statistics Schemas

class SubscriptionStats(BaseModel):
    organization_id: str
    total_subscriptions: int
    active_subscriptions: int
    subscriptions_by_type: dict[str, int]
    subscriptions_by_resource: dict[str, int]
    subscriptions_by_channel: dict[str, int]
    notifications_sent_today: int
    notifications_sent_this_week: int
    notifications_sent_this_month: int
    open_rate: float
    click_rate: float
    unsubscribe_rate: float


# Constants

SUBSCRIPTION_TYPE_LABELS: dict[SubscriptionType, str] = {
    SubscriptionType.INSTANT: "Instant Notifications",
    SubscriptionType.DAILY_DIGEST: "Daily Digest",
    SubscriptionType.WEEKLY_DIGEST: "Weekly Digest",
    SubscriptionType.MONTHLY_DIGEST: "Monthly Digest",
    SubscriptionType.SCHEDULED: "Scheduled",
    SubscriptionType.ON_REFRESH: "On Data Refresh",
    SubscriptionType.ON_THRESHOLD: "On Threshold",
}

SUBSCRIPTION_STATUS_LABELS: dict[SubscriptionStatus, str] = {
    SubscriptionStatus.ACTIVE: "Active",
    SubscriptionStatus.PAUSED: "Paused",
    SubscriptionStatus.EXPIRED: "Expired",
    SubscriptionStatus.CANCELLED: "Cancelled",
}

RESOURCE_TYPE_LABELS: dict[ResourceType, str] = {
    ResourceType.DASHBOARD: "Dashboard",
    ResourceType.CHART: "Chart",
    ResourceType.REPORT: "Report",
    ResourceType.KPI: "KPI",
    ResourceType.QUERY: "Query",
    ResourceType.ALERT: "Alert",
}

NOTIFICATION_CHANNEL_LABELS: dict[NotificationChannel, str] = {
    NotificationChannel.EMAIL: "Email",
    NotificationChannel.SLACK: "Slack",
    NotificationChannel.TEAMS: "Microsoft Teams",
    NotificationChannel.PUSH: "Push Notification",
    NotificationChannel.IN_APP: "In-App",
    NotificationChannel.SMS: "SMS",
    NotificationChannel.WEBHOOK: "Webhook",
}

THRESHOLD_OPERATOR_LABELS: dict[ThresholdOperator, str] = {
    ThresholdOperator.GREATER_THAN: "Greater than",
    ThresholdOperator.GREATER_THAN_OR_EQUAL: "Greater than or equal to",
    ThresholdOperator.LESS_THAN: "Less than",
    ThresholdOperator.LESS_THAN_OR_EQUAL: "Less than or equal to",
    ThresholdOperator.EQUAL: "Equal to",
    ThresholdOperator.NOT_EQUAL: "Not equal to",
    ThresholdOperator.CHANGE_BY: "Changes by",
    ThresholdOperator.CHANGE_BY_PERCENT: "Changes by %",
}


# Helper Functions

def evaluate_threshold(
    current_value: float,
    condition: ThresholdCondition,
    previous_value: Optional[float] = None,
) -> bool:
    """Evaluate if a threshold condition is met."""
    op = condition.operator
    threshold = condition.value

    if op == ThresholdOperator.GREATER_THAN:
        return current_value > threshold
    elif op == ThresholdOperator.GREATER_THAN_OR_EQUAL:
        return current_value >= threshold
    elif op == ThresholdOperator.LESS_THAN:
        return current_value < threshold
    elif op == ThresholdOperator.LESS_THAN_OR_EQUAL:
        return current_value <= threshold
    elif op == ThresholdOperator.EQUAL:
        return current_value == threshold
    elif op == ThresholdOperator.NOT_EQUAL:
        return current_value != threshold
    elif op == ThresholdOperator.CHANGE_BY:
        if previous_value is None:
            return False
        return abs(current_value - previous_value) >= threshold
    elif op == ThresholdOperator.CHANGE_BY_PERCENT:
        if previous_value is None or previous_value == 0:
            return False
        change_percent = abs((current_value - previous_value) / previous_value) * 100
        return change_percent >= threshold

    return False


def format_threshold_description(condition: ThresholdCondition) -> str:
    """Format threshold condition as human-readable string."""
    op_label = THRESHOLD_OPERATOR_LABELS.get(condition.operator, str(condition.operator))

    if condition.operator in [ThresholdOperator.CHANGE_BY, ThresholdOperator.CHANGE_BY_PERCENT]:
        unit = "%" if condition.operator == ThresholdOperator.CHANGE_BY_PERCENT else ""
        return f"{condition.metric_name} {op_label.lower()} {condition.value}{unit}"

    return f"{condition.metric_name} is {op_label.lower()} {condition.value}"


def should_send_digest(
    preferences: GlobalNotificationPreferences,
    last_digest_time: Optional[datetime] = None,
) -> bool:
    """Check if it's time to send a digest based on preferences."""
    if not preferences.digest_enabled:
        return False

    if last_digest_time is None:
        return True

    from datetime import timedelta

    now = datetime.utcnow()
    freq = preferences.digest_frequency

    if freq == DigestFrequency.DAILY:
        return (now - last_digest_time) >= timedelta(days=1)
    elif freq == DigestFrequency.WEEKLY:
        return (now - last_digest_time) >= timedelta(weeks=1)
    elif freq == DigestFrequency.BIWEEKLY:
        return (now - last_digest_time) >= timedelta(weeks=2)
    elif freq == DigestFrequency.MONTHLY:
        return (now - last_digest_time) >= timedelta(days=30)

    return False


def is_in_quiet_hours(preferences: GlobalNotificationPreferences) -> bool:
    """Check if current time is within quiet hours."""
    if not preferences.quiet_hours_enabled:
        return False

    from datetime import datetime as dt

    now = dt.utcnow()
    # In production, convert to user's timezone
    current_time = now.strftime("%H:%M")

    start = preferences.quiet_hours_start
    end = preferences.quiet_hours_end

    # Handle overnight quiet hours (e.g., 22:00 to 08:00)
    if start > end:
        return current_time >= start or current_time < end
    else:
        return start <= current_time < end

File: frontend/src/types/report_subscription.ts
/**
 * Report Subscriptions Types
 *
 * TypeScript types for report subscription management including
 * subscriptions, notifications, digests, and user preferences.
 */

// Enums

export enum SubscriptionType {
  INSTANT = 'instant',
  DAILY_DIGEST = 'daily_digest',
  WEEKLY_DIGEST = 'weekly_digest',
  MONTHLY_DIGEST = 'monthly_digest',
  SCHEDULED = 'scheduled',
  ON_REFRESH = 'on_refresh',
  ON_THRESHOLD = 'on_threshold',
}

export enum SubscriptionStatus {
  ACTIVE = 'active',
  PAUSED = 'paused',
  EXPIRED = 'expired',
  CANCELLED = 'cancelled',
}

export enum ResourceType {
  DASHBOARD = 'dashboard',
  CHART = 'chart',
  REPORT = 'report',
  KPI = 'kpi',
  QUERY = 'query',
  ALERT = 'alert',
}

export enum NotificationChannel {
  EMAIL = 'email',
  SLACK = 'slack',
  TEAMS = 'teams',
  PUSH = 'push',
  IN_APP = 'in_app',
  SMS = 'sms',
  WEBHOOK = 'webhook',
}

export enum DigestFrequency {
  DAILY = 'daily',
  WEEKLY = 'weekly',
  BIWEEKLY = 'biweekly',
  MONTHLY = 'monthly',
}

export enum ThresholdOperator {
  GREATER_THAN = 'gt',
  GREATER_THAN_OR_EQUAL = 'gte',
  LESS_THAN = 'lt',
  LESS_THAN_OR_EQUAL = 'lte',
  EQUAL = 'eq',
  NOT_EQUAL = 'neq',
  CHANGE_BY = 'change_by',
  CHANGE_BY_PERCENT = 'change_by_percent',
}

// Notification Configuration Interfaces

export interface EmailNotificationConfig {
  email_address?: string;
  include_preview: boolean;
  include_attachment: boolean;
  attachment_format: string;
  subject_prefix: string;
}

export interface SlackNotificationConfig {
  channel_id?: string;
  dm_user: boolean;
  include_preview_image: boolean;
  mention_on_threshold: boolean;
}

export interface TeamsNotificationConfig {
  channel_webhook?: string;
  include_preview_image: boolean;
}

export interface PushNotificationConfig {
  send_to_mobile: boolean;
  send_to_desktop: boolean;
  sound_enabled: boolean;
}

export interface WebhookNotificationConfig {
  url: string;
  method: string;
  headers: Record<string, string>;
  include_data: boolean;
}

export interface NotificationConfig {
  channel: NotificationChannel;
  email_config?: EmailNotificationConfig;
  slack_config?: SlackNotificationConfig;
  teams_config?: TeamsNotificationConfig;
  push_config?: PushNotificationConfig;
  webhook_config?: WebhookNotificationConfig;
  enabled: boolean;
}

// Threshold Configuration

export interface ThresholdCondition {
  metric_name: string;
  operator: ThresholdOperator;
  value: number;
  secondary_value?: number;
  comparison_period?: string;
}

// Schedule Configuration

export interface SubscriptionSchedule {
  days_of_week: string[];
  time_of_day: string;
  timezone: string;
}

// Digest Configuration

export interface DigestConfig {
  frequency: DigestFrequency;
  day_of_week?: string;
  day_of_month?: number;
  time_of_day: string;
  timezone: string;
  include_summary: boolean;
  include_highlights: boolean;
  max_items: number;
  group_by_resource: boolean;
}

// Subscription Interfaces

export interface Subscription {
  id: string;
  user_id: string;
  organization_id?: string;
  resource_type: ResourceType;
  resource_id: string;
  resource_name?: string;
  subscription_type: SubscriptionType;
  notification_channels: NotificationConfig[];
  status: SubscriptionStatus;
  is_active: boolean;
  schedule?: SubscriptionSchedule;
  digest_config?: DigestConfig;
  threshold_conditions: ThresholdCondition[];
  filters: Record<string, unknown>;
  last_notified_at?: string;
  notification_count: number;
  expires_at?: string;
  created_at: string;
  updated_at: string;
}

export interface SubscriptionCreate {
  resource_type: ResourceType;
  resource_id: string;
  resource_name?: string;
  subscription_type: SubscriptionType;
  notification_channels: NotificationConfig[];
  schedule?: SubscriptionSchedule;
  digest_config?: DigestConfig;
  threshold_conditions?: ThresholdCondition[];
  filters?: Record<string, unknown>;
  is_active?: boolean;
  expires_at?: string;
}

export interface SubscriptionUpdate {
  resource_name?: string;
  subscription_type?: SubscriptionType;
  notification_channels?: NotificationConfig[];
  schedule?: SubscriptionSchedule;
  digest_config?: DigestConfig;
  threshold_conditions?: ThresholdCondition[];
  filters?: Record<string, unknown>;
  is_active?: boolean;
  status?: SubscriptionStatus;
  expires_at?: string;
}

export interface SubscriptionListResponse {
  subscriptions: Subscription[];
  total: number;
}

// Notification Interfaces

export interface NotificationDelivery {
  channel: NotificationChannel;
  success: boolean;
  delivered_at?: string;
  error?: string;
}

export interface SubscriptionNotification {
  id: string;
  subscription_id: string;
  user_id: string;
  resource_type: ResourceType;
  resource_id: string;
  resource_name?: string;
  notification_type: string;
  title: string;
  message: string;
  data: Record<string, unknown>;
  deliveries: NotificationDelivery[];
  read: boolean;
  read_at?: string;
  clicked: boolean;
  clicked_at?: string;
  created_at: string;
}

export interface NotificationListResponse {
  notifications: SubscriptionNotification[];
  total: number;
  unread_count: number;
}

// Digest Interfaces

export interface DigestItem {
  resource_type: ResourceType;
  resource_id: string;
  resource_name: string;
  summary: string;
  changes: string[];
  metrics: Record<string, unknown>;
  timestamp: string;
}

export interface Digest {
  id: string;
  user_id: string;
  digest_type: DigestFrequency;
  period_start: string;
  period_end: string;
  items: DigestItem[];
  total_items: number;
  highlights: string[];
  sent_at?: string;
  created_at: string;
}

export interface DigestListResponse {
  digests: Digest[];
  total: number;
}

// User Preferences Interfaces

export interface GlobalNotificationPreferences {
  email_enabled: boolean;
  push_enabled: boolean;
  in_app_enabled: boolean;
  slack_enabled: boolean;
  teams_enabled: boolean;
  quiet_hours_enabled: boolean;
  quiet_hours_start: string;
  quiet_hours_end: string;
  timezone: string;
  digest_enabled: boolean;
  digest_frequency: DigestFrequency;
  digest_time: string;
  unsubscribe_from_marketing: boolean;
}

export interface UserSubscriptionPreferences {
  user_id: string;
  global_preferences: GlobalNotificationPreferences;
  channel_preferences: Record<string, boolean>;
  resource_preferences: Record<string, boolean>;
  muted_resources: string[];
  updated_at: string;
}

// Statistics Interfaces

export interface SubscriptionStats {
  organization_id: string;
  total_subscriptions: number;
  active_subscriptions: number;
  subscriptions_by_type: Record<string, number>;
  subscriptions_by_resource: Record<string, number>;
  subscriptions_by_channel: Record<string, number>;
  notifications_sent_today: number;
  notifications_sent_this_week: number;
  notifications_sent_this_month: number;
  open_rate: number;
  click_rate: number;
  unsubscribe_rate: number;
}

// Constants

export const SUBSCRIPTION_TYPE_LABELS: Record<SubscriptionType, string> = {
  [SubscriptionType.INSTANT]: 'Instant Notifications',
  [SubscriptionType.DAILY_DIGEST]: 'Daily Digest',
  [SubscriptionType.WEEKLY_DIGEST]: 'Weekly Digest',
  [SubscriptionType.MONTHLY_DIGEST]: 'Monthly Digest',
  [SubscriptionType.SCHEDULED]: 'Scheduled',
  [SubscriptionType.ON_REFRESH]: 'On Data Refresh',
  [SubscriptionType.ON_THRESHOLD]: 'On Threshold',
};

export const SUBSCRIPTION_STATUS_LABELS: Record<SubscriptionStatus, string> = {
  [SubscriptionStatus.ACTIVE]: 'Active',
  [SubscriptionStatus.PAUSED]: 'Paused',
  [SubscriptionStatus.EXPIRED]: 'Expired',
  [SubscriptionStatus.CANCELLED]: 'Cancelled',
};

export const RESOURCE_TYPE_LABELS: Record<ResourceType, string> = {
  [ResourceType.DASHBOARD]: 'Dashboard',
  [ResourceType.CHART]: 'Chart',
  [ResourceType.REPORT]: 'Report',
  [ResourceType.KPI]: 'KPI',
  [ResourceType.QUERY]: 'Query',
  [ResourceType.ALERT]: 'Alert',
};

export const NOTIFICATION_CHANNEL_LABELS: Record<NotificationChannel, string> = {
  [NotificationChannel.EMAIL]: 'Email',
  [NotificationChannel.SLACK]: 'Slack',
  [NotificationChannel.TEAMS]: 'Microsoft Teams',
  [NotificationChannel.PUSH]: 'Push Notification',
  [NotificationChannel.IN_APP]: 'In-App',
  [NotificationChannel.SMS]: 'SMS',
  [NotificationChannel.WEBHOOK]: 'Webhook',
};

export const THRESHOLD_OPERATOR_LABELS: Record<ThresholdOperator, string> = {
  [ThresholdOperator.GREATER_THAN]: 'Greater than',
  [ThresholdOperator.GREATER_THAN_OR_EQUAL]: 'Greater than or equal to',
  [ThresholdOperator.LESS_THAN]: 'Less than',
  [ThresholdOperator.LESS_THAN_OR_EQUAL]: 'Less than or equal to',
  [ThresholdOperator.EQUAL]: 'Equal to',
  [ThresholdOperator.NOT_EQUAL]: 'Not equal to',
  [ThresholdOperator.CHANGE_BY]: 'Changes by',
  [ThresholdOperator.CHANGE_BY_PERCENT]: 'Changes by %',
};

export const DIGEST_FREQUENCY_LABELS: Record<DigestFrequency, string> = {
  [DigestFrequency.DAILY]: 'Daily',
  [DigestFrequency.WEEKLY]: 'Weekly',
  [DigestFrequency.BIWEEKLY]: 'Bi-Weekly',
  [DigestFrequency.MONTHLY]: 'Monthly',
};

export const SUBSCRIPTION_STATUS_COLORS: Record<SubscriptionStatus, string> = {
  [SubscriptionStatus.ACTIVE]: 'green',
  [SubscriptionStatus.PAUSED]: 'yellow',
  [SubscriptionStatus.EXPIRED]: 'gray',
  [SubscriptionStatus.CANCELLED]: 'red',
};

// Helper Functions

export function getDefaultGlobalPreferences(): GlobalNotificationPreferences {
  return {
    email_enabled: true,
    push_enabled: true,
    in_app_enabled: true,
    slack_enabled: false,
    teams_enabled: false,
    quiet_hours_enabled: false,
    quiet_hours_start: '22:00',
    quiet_hours_end: '08:00',
    timezone: 'UTC',
    digest_enabled: true,
    digest_frequency: DigestFrequency.DAILY,
    digest_time: '09:00',
    unsubscribe_from_marketing: false,
  };
}

export function getDefaultEmailNotificationConfig(): NotificationConfig {
  return {
    channel: NotificationChannel.EMAIL,
    email_config: {
      include_preview: true,
      include_attachment: false,
      attachment_format: 'pdf',
      subject_prefix: '[DataViz]',
    },
    enabled: true,
  };
}

export function getDefaultDigestConfig(): DigestConfig {
  return {
    frequency: DigestFrequency.DAILY,
    time_of_day: '09:00',
    timezone: 'UTC',
    include_summary: true,
    include_highlights: true,
    max_items: 10,
    group_by_resource: true,
  };
}

export function isSubscriptionActive(subscription: Subscription): boolean {
  return subscription.status === SubscriptionStatus.ACTIVE && subscription.is_active;
}

export function hasUnreadNotifications(response: NotificationListResponse): boolean {
  return response.unread_count > 0;
}

export function getChannelIcon(channel: NotificationChannel): string {
  const icons: Record<NotificationChannel, string> = {
    [NotificationChannel.EMAIL]: 'mail',
    [NotificationChannel.SLACK]: 'message-square',
    [NotificationChannel.TEAMS]: 'users',
    [NotificationChannel.PUSH]: 'bell',
    [NotificationChannel.IN_APP]: 'inbox',
    [NotificationChannel.SMS]: 'smartphone',
    [NotificationChannel.WEBHOOK]: 'globe',
  };
  return icons[channel] || 'bell';
}

export function getResourceTypeIcon(resourceType: ResourceType): string {
  const icons: Record<ResourceType, string> = {
    [ResourceType.DASHBOARD]: 'layout-dashboard',
    [ResourceType.CHART]: 'bar-chart',
    [ResourceType.REPORT]: 'file-text',
    [ResourceType.KPI]: 'trending-up',
    [ResourceType.QUERY]: 'database',
    [ResourceType.ALERT]: 'alert-circle',
  };
  return icons[resourceType] || 'file';
}

export function getSubscriptionTypeIcon(type: SubscriptionType): string {
  const icons: Record<SubscriptionType, string> = {
    [SubscriptionType.INSTANT]: 'zap',
    [SubscriptionType.DAILY_DIGEST]: 'calendar',
    [SubscriptionType.WEEKLY_DIGEST]: 'calendar-days',
    [SubscriptionType.MONTHLY_DIGEST]: 'calendar-range',
    [SubscriptionType.SCHEDULED]: 'clock',
    [SubscriptionType.ON_REFRESH]: 'refresh-cw',
    [SubscriptionType.ON_THRESHOLD]: 'alert-triangle',
  };
  return icons[type] || 'bell';
}

export function formatThresholdDescription(condition: ThresholdCondition): string {
  const opLabel = THRESHOLD_OPERATOR_LABELS[condition.operator] || condition.operator;

  if (
    condition.operator === ThresholdOperator.CHANGE_BY ||
    condition.operator === ThresholdOperator.CHANGE_BY_PERCENT
  ) {
    const unit = condition.operator === ThresholdOperator.CHANGE_BY_PERCENT ? '%' : '';
    return `${condition.metric_name} ${opLabel.toLowerCase()} ${condition.value}${unit}`;
  }

  return `${condition.metric_name} is ${opLabel.toLowerCase()} ${condition.value}`;
}

export function getSubscriptionDescription(subscription: Subscription): string {
  const type = SUBSCRIPTION_TYPE_LABELS[subscription.subscription_type];
  const resource = RESOURCE_TYPE_LABELS[subscription.resource_type];
  const channels = subscription.notification_channels
    .filter(c => c.enabled)
    .map(c => NOTIFICATION_CHANNEL_LABELS[c.channel])
    .join(', ');

  return `${type} for ${resource} via ${channels || 'no channels'}`;
}

export function canPauseSubscription(subscription: Subscription): boolean {
  return subscription.status === SubscriptionStatus.ACTIVE;
}

export function canResumeSubscription(subscription: Subscription): boolean {
  return subscription.status === SubscriptionStatus.PAUSED;
}

export function getEnabledChannels(subscription: Subscription): NotificationChannel[] {
  return subscription.notification_channels
    .filter(c => c.enabled)
    .map(c => c.channel);
}
